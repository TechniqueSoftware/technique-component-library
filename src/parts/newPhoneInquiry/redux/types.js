export const key = 'newPhoneInquiry'

const SUBMIT = {
  SEND: `${key}/SUBMIT/SEND`,
  ERROR: `${key}/SUBMIT/ERROR`,

  RECEIVE_SUCCESS: `${key}/SUBMIT/RECEIVE_SUCCESS`,
  RECEIVE_ERROR: `${key}/SUBMIT/RECEIVE_ERROR`,

  RECEIVE_APPT_SUCCESS: `${key}/SUBMIT/RECEIVE_APPT_SUCCESS`,
  RECEIVE_APPT_ERROR: `${key}/SUBMIT/RECEIVE_APPT_ERROR`
}

const DATA = {
  SET: `${key}/DATA/SET`,
  TOGGLE_APPOINTMENT: `${key}/DATA/TOGGLE_APPOINTMENT`,

  CLEAR: `${key}/DATA/CLEAR`,

  REQUEST_USER_DATA: `${key}/DATA/REQUEST_USER_DATA`,
  RECEIVE_USER_DATA: `${key}/DATA/RECEIVE_USER_DATA`,

  REQUEST_EMPLOYEES_DATA: `${key}/DATA/REQUEST_EMPLOYEES_DATA`,
  RECEIVE_EMPLOYEES_DATA: `${key}/DATA/RECEIVE_EMPLOYEES_DATA`,

  REQUEST_SCRIPT_DATA: `${key}/DATA/REQUEST_SCRIPT_DATA`,
  RECEIVE_SCRIPT_DATA: `${key}/DATA/RECEIVE_SCRIPT_DATA`,

  SELECT_LOCATION: `${key}/DATA/SELECT_LOCATION`
}

export default {
  SUBMIT,
  DATA
}
