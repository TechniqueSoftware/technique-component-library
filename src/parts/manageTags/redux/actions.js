import TYPES from './types'

const DATA_ACTIONS = {
  SET: payload => ({ type: TYPES.DATA.SET, payload }),

  REQUEST_TAGS: payload => ({ type: TYPES.DATA.REQUEST_TAGS, payload }),
  SAVE_TAGS: payload => ({ type: TYPES.DATA.SAVE_TAGS, payload }),
  DELETE_TAGS: payload => ({ type: TYPES.DATA.DELETE_TAGS, payload })
}

export default {
  DATA: DATA_ACTIONS
}
