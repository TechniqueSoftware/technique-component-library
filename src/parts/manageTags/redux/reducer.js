import { produce } from 'immer'

import TYPES from './types'

export const initialState = {
  tagsList: []
}

const reducer = (state = initialState, { type, payload }) => produce(state, draft => {
  switch (type) {
    case TYPES.DATA.SET:
      draft.tagsList = payload
      break
  }
})

export default reducer
