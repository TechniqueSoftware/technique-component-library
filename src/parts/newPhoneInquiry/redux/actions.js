import TYPES from './types'

const SUBMIT_ACTIONS = {
  SEND: () => ({ type: TYPES.SUBMIT.SEND }),
  ERROR: payload => ({ type: TYPES.SUBMIT.ERROR, payload }),

  RECEIVE_SUCCESS: payload => ({ type: TYPES.SUBMIT.RECEIVE_SUCCESS, payload }),
  RECEIVE_ERROR: payload => ({ type: TYPES.SUBMIT.RECEIVE_ERROR, payload }),

  RECEIVE_APPT_SUCCESS: payload => ({ type: TYPES.SUBMIT.RECEIVE_APPT_SUCCESS, payload }),
  RECEIVE_APPT_ERROR: payload => ({ type: TYPES.SUBMIT.RECEIVE_APPT_ERROR, payload })
}

const DATA_ACTIONS = {
  SET: payload => ({ type: TYPES.DATA.SET, payload }),
  TOGGLE_APPOINTMENT: payload => ({ type: TYPES.DATA.TOGGLE_APPOINTMENT, payload }),

  CLEAR: () => ({ type: TYPES.DATA.CLEAR }),
  REQUEST_USER_DATA: payload => ({ type: TYPES.DATA.REQUEST_USER_DATA, payload }),
  RECEIVE_USER_DATA: payload => ({ type: TYPES.DATA.RECEIVE_USER_DATA, payload }),

  REQUEST_EMPLOYEES_DATA: payload => ({ type: TYPES.DATA.REQUEST_EMPLOYEES_DATA, payload }),
  RECEIVE_EMPLOYEES_DATA: payload => ({ type: TYPES.DATA.RECEIVE_EMPLOYEES_DATA, payload }),

  REQUEST_SCRIPT_DATA: payload => ({ type: TYPES.DATA.REQUEST_SCRIPT_DATA, payload }),
  RECEIVE_SCRIPT_DATA: payload => ({ type: TYPES.DATA.RECEIVE_SCRIPT_DATA, payload }),

  SELECT_LOCATION: payload => ({ type: TYPES.DATA.SELECT_LOCATION, payload })
}

export default {
  SUBMIT: SUBMIT_ACTIONS,
  DATA: DATA_ACTIONS
}
