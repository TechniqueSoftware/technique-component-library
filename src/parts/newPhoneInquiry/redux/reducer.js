import { produce } from 'immer'

import moment from 'moment-timezone'
import TYPES from './types'

export const initialState = {
  errors: [],
  appointmentEnabled: false,
  form: {
    locationSelect: {
      type: 'select',
      label: 'Location',
      value: {
        name: '',
        value: '',
        timeZone: ''
      },
      error: '',
      required: true,
      relatedField: null,
      validated: false
    },
    firstName: {
      type: 'text',
      label: 'First Name',
      value: '',
      error: '',
      required: true,
      relatedField: null,
      validated: false
    },
    lastName: {
      type: 'text',
      label: 'Last Name',
      value: '',
      error: '',
      required: true,
      relatedField: null,
      validated: false
    },
    email: {
      type: 'email',
      label: 'Email',
      value: '',
      error: '',
      required: false,
      relatedField: 'mobilePhone',
      validated: false
    },
    mobilePhone: {
      type: 'phone',
      label: 'Phone',
      value: '',
      error: '',
      required: false,
      relatedField: 'email',
      validated: false
    },
    contact: {
      type: 'select',
      label: 'With',
      value: {
        name: '',
        value: '',
        availability: []
      },
      error: '',
      relatedField: null,
      validated: false
    },
    date: {
      type: 'date',
      label: 'Date',
      value: moment().format('YYYY-MM-DD'),
      error: '',
      relatedField: null,
      validated: false
    },
    time: {
      type: 'select',
      label: 'Time',
      value: {
        name: '',
        value: ''
      },
      error: '',
      relatedField: null,
      validated: false
    }
  },
  tiScript: {
    value: 'Loading...'
  },
  response: {}
}

const reducer = (state = initialState, { type, payload }) => produce(state, draft => {
  switch (type) {
    case TYPES.SUBMIT.SEND:
      draft.loading = true
      break
    case TYPES.SUBMIT.ERROR:
      draft.response = payload
      break
    case TYPES.SUBMIT.RECEIVE_SUCCESS:
      draft.loading = false
      draft.response = payload
      break
    case TYPES.SUBMIT.RECEIVE_ERROR:
      draft.loading = false
      draft.response = payload
      break

    case TYPES.DATA.SET:
      draft.form[payload.name] = { ...draft.form[payload.name], ...payload.data }
      break
    case TYPES.DATA.TOGGLE_APPOINTMENT:
      draft.appointmentEnabled = payload
      break

    case TYPES.DATA.CLEAR:
      draft.response = {}
      draft.appointmentEnabled = false
      draft.form = {
        ...initialState.form,
        locationSelect: draft.form.locationSelect,
        contact: draft.form.contact,
        date: draft.form.date,
        time: draft.form.time
      }
      break
    case TYPES.DATA.RECEIVE_USER_DATA:
      draft.response.user = payload
      break
    case TYPES.DATA.RECEIVE_EMPLOYEES_DATA:
      draft.locationEmployees = payload
      break
    case TYPES.DATA.RECEIVE_SCRIPT_DATA:
      draft.tiScript.value = payload
      break
  }
})

export default reducer
