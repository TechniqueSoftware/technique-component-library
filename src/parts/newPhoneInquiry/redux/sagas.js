import { call, put, select, takeEvery, takeLatest } from 'redux-saga/effects'
import axios from '@axios'
import moment from 'moment-timezone'

import { GLOBAL_SELECTOR } from '@redux/global'

import SELECTORS from './selectors'
import ACTIONS from './actions'
import TYPES from './types'

const formatAppointmentTime = (time, timeZone) => {
  if (!time.value?.value) return [null, null]
  const utcStart = moment.tz(time.value.value, timeZone).utc().format()
  const date = new Date(utcStart)
  const utcEnd = new Date(date.getTime() + 30 * 60000).toISOString()
  return [utcStart, utcEnd]
}

const formatFormData = formData => Object.keys(formData).reduce((obj, key) => {
  if (!['locationSelect'].includes(key)) return { ...obj, [key]: formData[key].value }
  return obj
}, {})

const simplifyErrorStructure = err => {
  if (err.status === 400) return { status: err.status, code: err.code, user: err.user }
  if (err.status === 'BAD_REQUEST') {
    let message = ''
    if (err.messages.code.includes('DATA TOO LONG')) {
      if (err.messages.code.includes('FIRST_NAME')) message = 'First name cannot exceed 50 characters'
      if (err.messages.code.includes('LAST_NAME')) message = 'Last name cannot exceed 50 characters'
    }
    return { status: 400, code: err.status, error: err.error, message }
  }
}

export const apiCalls = {
  submit: formBody => axios.post('/users', formBody),
  getUserData: userId => axios.get(`/users/${userId}?limited=true`),
  getLocationData: locationId => axios.get(`/locations/?locationId=${locationId}`),
  getScriptData: locationId => axios.get(`/settings/ti-script/${locationId}`),
  getEmployeesAvailability: ({ locationId, params }) => axios.get(`/locations/${locationId}/availability?eventTypeId=0&eventTypeId=4`, { params }), // eslint-disable-line max-len
  getOwnAvailability: ({ userId, params }) => axios.get(`/users/${userId}/availability?eventTypeId=0&eventTypeId=4`, { params }), // eslint-disable-line max-len
  createAppointmentEvent: appointmentBody => axios.post('/events', appointmentBody)
}

export function* watchGetUserData (action) {
  try {
    if (!action?.payload) throw new Error('There is no user')
    const userData = yield call(apiCalls.getUserData, action.payload)
    yield put(ACTIONS.DATA.RECEIVE_USER_DATA(userData))
  } catch (err) {
    console.error(err) // eslint-disable-line no-console
  }
}

export function* formSubmission () {
  try {
    const user = yield select(GLOBAL_SELECTOR.USER)
    const selectedLocation = yield select(SELECTORS.SELECTED_LOCATION)
    const formData = yield select(SELECTORS.FORM_DATA)
    const appointment = yield select(SELECTORS.APPOINTMENT)
    const appointmentEnabled = yield select(SELECTORS.APPOINTMENT_ENABLED)

    const updatedFormData = appointmentEnabled
      ? formData
      : {
        ...formData,
        contact: {
          value: {
            value: user.userId,
            name: `${user.firstName} ${user.lastName}`
          }
        }
      }

    const userBody = {
      ...formatFormData(updatedFormData),
      locationId: selectedLocation?.value?.value,
      salespersonId: appointmentEnabled ? appointment?.contact?.value.value : user.userId,
      clubId: user.clubId,
      gender: 'M',
      role: { id: 99, name: 'Prospect' },
      source: 'T.I.',
      userOrigin: {
        originId: 2,
        originName: 'T.I.'
      },
      sendResponseEmails: false,
      smsOptOut: true,
      emailOptOut: false
    }

    const data = yield call(apiCalls.submit, JSON.stringify(userBody))
    data.status = 202
    data.code = 'SUCCESS'

    if (appointmentEnabled) {
      const { timeZone, value } = selectedLocation.value
      const [startTime, endTime] = formatAppointmentTime(appointment.time, timeZone)
      const appointmentBody = {
        clubId: user.clubId,
        locationId: value,
        eventTypeId: 4,
        eventOwnerId: appointment?.contact?.value.value,
        startTime,
        endTime,
        subject: null,
        createdById: user.userId,
        remindAttendee: true,
        remindOwner: true,
        bookingSource: 'Club OS T.I. Form',
        attendee: [
          {
            attendeeId: data.userId,
            attendeeStatus: 'A',
            visitType: 'A'
          }
        ]
      }

      const appointmentResponse = yield call(apiCalls.createAppointmentEvent, JSON.stringify(appointmentBody))

      yield put(ACTIONS.SUBMIT.RECEIVE_APPT_SUCCESS(appointmentResponse))
      yield put(ACTIONS.DATA.CLEAR())
    }
    yield put(ACTIONS.SUBMIT.RECEIVE_SUCCESS(data))
  } catch (err) {
    const modifiedError = simplifyErrorStructure(err)

    yield put(ACTIONS.SUBMIT.RECEIVE_ERROR(modifiedError))
    yield put(ACTIONS.SUBMIT.RECEIVE_APPT_ERROR(modifiedError))
  }
}

export function* selectLocation ({ payload }) {
  try {
    const { name, error, validated, value } = payload
    const locationId = value.value
    const dataResponse = yield call(apiCalls.getLocationData, locationId)
    const timeZone = dataResponse?.content[0]?.timeZone
    yield put(
      ACTIONS.DATA.SET({
        name,
        data: { value: { ...value, timeZone }, error, validated }
      })
    )
  } catch (err) {
    return null
  }
}

export function* watchScriptData () {
  try {
    const { value } = yield select(SELECTORS.SELECTED_LOCATION)
    const locationId = value.value

    if (locationId) {
      const scriptData = yield call(apiCalls.getScriptData, locationId)
      yield put(ACTIONS.DATA.RECEIVE_SCRIPT_DATA(scriptData?.script))
    }
  } catch (err) {
    console.error(err) // eslint-disable-line no-console
  }
}

export function* getEmployees () {
  try {
    const selectedLocation = yield select(SELECTORS.SELECTED_LOCATION)
    const { date } = yield select(SELECTORS.APPOINTMENT)
    const { userId } = yield select(GLOBAL_SELECTOR.USER)
    const { value: { value, timeZone } } = selectedLocation
    const locationId = value
    const offset = moment.tz(timeZone).format('Z')
    const params = {
      availabilityDuration: 30,
      startTimeStart: `${date.value}T00:00:00.${offset}`,
      startTimeEnd: `${date.value}T23:45:00.${offset}`,
      multiBookingEventTypeId: 4
    }
    const employeesParams = {
      ...params,
      status: 'available'
    }
    const ownParams = {
      ...params,
      locationId,
      applyBookingThreshold: false
    }
    const employeesRes = yield call(apiCalls.getEmployeesAvailability, { locationId, params: employeesParams })
    const ownAvailabilityRes = yield call(apiCalls.getOwnAvailability, { userId, params: ownParams })
    const users = employeesRes.users.reduce((acc, user) => {
      if (user.userId === userId) return acc.concat(ownAvailabilityRes)
      if (!user.availability.length) return acc
      return acc.concat(user)
    }, [])

    const sortedUsers = [...users].sort((a, b) => a.firstName.localeCompare(b.firstName))
    yield put(ACTIONS.DATA.RECEIVE_EMPLOYEES_DATA(sortedUsers))
  } catch (err) {
    console.error(err) // eslint-disable-line no-console
  }
}

export default function* saga () {
  yield takeEvery(TYPES.DATA.REQUEST_USER_DATA, watchGetUserData)
  yield takeLatest(TYPES.SUBMIT.SEND, formSubmission)
  yield takeLatest(TYPES.DATA.SELECT_LOCATION, selectLocation)
  yield takeLatest(TYPES.DATA.REQUEST_SCRIPT_DATA, watchScriptData)
  yield takeLatest(TYPES.DATA.REQUEST_EMPLOYEES_DATA, getEmployees)
}
