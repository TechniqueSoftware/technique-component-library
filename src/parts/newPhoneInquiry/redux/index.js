import { createManager } from '@redux/injector'

import reducer from './reducer'
import { key } from './types'
import saga from './sagas'

export { default as REDUCER } from './reducer'
export { default as ACTIONS } from './actions'
export { default as SELECTORS } from './selectors'
export { key, default as TYPES } from './types'

export const Manager = createManager({
  name: 'NewPhoneInquiryManager',
  key,
  reducer,
  saga
})
