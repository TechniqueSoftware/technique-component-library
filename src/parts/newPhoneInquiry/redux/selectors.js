import { createSelector } from 'reselect'
import { get } from 'lodash'

import { key } from './types'
import { initialState } from './reducer'

const reducer = state => get(state, key, initialState)

const getFormResponse = createSelector(
  reducer, state => get(state, 'response', {})
)

const getForm = createSelector(
  reducer, state => get(state, 'form', initialState.form)
)

const getConstants = createSelector(
  reducer, state => get(state, 'constants', initialState.formConstants)
)

const getScriptData = createSelector(
  reducer, state => get(state, 'tiScript[value]', '')
)

const getEmployees = createSelector(
  reducer, state => get(state, 'locationEmployees', [])
)

const getAppointment = createSelector(reducer, state => {
  const form = get(state, 'form', initialState.form)
  return { contact: form.contact, date: form.date, time: form.time }
})

const getAppointmentEnabled = createSelector(
  reducer, state => get(state, 'appointmentEnabled', false)
)

const getSelectedLocation = createSelector(
  reducer, state => get(state, 'form[locationSelect]', initialState.form.locationSelect)
)

export default {
  FORM_RESPONSE: getFormResponse,
  FORM_DATA: getForm,
  FORM_CONSTANTS: getConstants,
  SCRIPT: getScriptData,
  EMPLOYEES: getEmployees,
  APPOINTMENT: getAppointment,
  APPOINTMENT_ENABLED: getAppointmentEnabled,
  SELECTED_LOCATION: getSelectedLocation
}
