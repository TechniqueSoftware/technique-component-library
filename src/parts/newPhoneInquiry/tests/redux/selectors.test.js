import { SELECTORS } from '../../redux'

import { state } from './mock'

describe('SELECTORS', () => {
  it('get form', () => {
    expect(SELECTORS.FORM_DATA(state)).toMatchSnapshot()
  })
  it('get form response', () => {
    expect(SELECTORS.FORM_RESPONSE(state)).toMatchSnapshot()
  })
})
