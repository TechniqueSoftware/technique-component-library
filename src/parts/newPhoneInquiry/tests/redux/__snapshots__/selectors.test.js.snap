// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SELECTORS get form 1`] = `
Object {
  "contact": Object {
    "error": "",
    "label": "With",
    "relatedField": null,
    "required": false,
    "type": "select",
    "validated": true,
    "value": Object {
      "availability": Array [],
      "name": "location managerfive",
      "value": 136951500,
    },
  },
  "date": Object {
    "error": "",
    "label": "Date",
    "relatedField": null,
    "required": false,
    "type": "date",
    "validated": true,
    "value": "2022-10-17",
  },
  "email": Object {
    "error": "",
    "label": "Email",
    "relatedField": "mobilePhone",
    "required": false,
    "type": "email",
    "validated": true,
    "value": "testappttime120@aol.com",
  },
  "firstName": Object {
    "error": "",
    "label": "First Name",
    "relatedField": null,
    "required": true,
    "type": "text",
    "validated": true,
    "value": "Test",
  },
  "lastName": Object {
    "error": "",
    "label": "Last Name",
    "relatedField": null,
    "required": true,
    "type": "text",
    "validated": true,
    "value": "ApptTime120",
  },
  "locationSelect": Object {
    "error": "",
    "label": "Location",
    "relatedField": null,
    "required": true,
    "type": "select",
    "validated": true,
    "value": Object {
      "name": "Philadelphia, PA",
      "value": 4559,
    },
  },
  "mobilePhone": Object {
    "error": "",
    "label": "Phone",
    "relatedField": "email",
    "required": false,
    "type": "phone",
    "validated": true,
    "value": "",
  },
  "time": Object {
    "error": "",
    "label": "Time",
    "relatedField": null,
    "required": false,
    "type": "select",
    "validated": true,
    "value": Object {
      "name": "12:00 AM",
      "value": "2022-10-17T16:00:00Z",
    },
  },
}
`;

exports[`SELECTORS get form response 1`] = `Object {}`;
